File: options-results/options_results.cpp
#include <iostream>
#include <optional>
#include <expected>
#include <string>
#include <cstdint>

std::optional<uint32_t> find(uint32_t id) {
    if (id > 0) return id;
    return std::nullopt;
}

std::expected<int32_t, std::string> divide(int32_t a, int32_t b) {
    if (b == 0) return std::unexpected("division by zero");
    return a / b;
}

int main() {
    // Option example
    auto id = 42;
    if (auto found_id = find(id)) {
        std::cout << "Found: " << *found_id << '\n';
    } else {
        std::cout << "Not found\n";
    }

    // Check non-existent id
    if (auto found_id = find(0)) {
        std::cout << "Found: " << *found_id << '\n';
    } else {
        std::cout << "Not found\n";
    }

    // Result example
    auto result1 = divide(10, 2);
    if (result1) {
        std::cout << "10/2 = " << *result1 << '\n';
    } else {
        std::cout << "Error: " << result1.error() << '\n';
    }

    auto result2 = divide(5, 0);
    if (result2) {
        std::cout << "5/0 = " << *result2 << '\n';
    } else {
        std::cout << "Error: " << result2.error() << '\n';
    }

    return 0;
}

File: options-results/options_results.rs
// Implementation of options and results from README.md

fn find(id: u32) -> Option<u32> {
    if id > 0 { Some(id) } else { None }
}

fn divide(a: i32, b: i32) -> Result<i32, &'static str> {
    if b == 0 { Err("division by zero") } else { Ok(a / b) }
}

fn main() {
    // Option example
    let id = 42;
    match find(id) {
        Some(found_id) => println!("Found: {}", found_id),
        None => println!("Not found"),
    }

    // Also works with if let
    if let Some(found_id) = find(0) {
        println!("Found: {}", found_id);
    } else {
        println!("Not found");
    }

    // Result example
    match divide(10, 2) {
        Ok(result) => println!("10/2 = {}", result),
        Err(e) => println!("Error: {}", e),
    }

    match divide(5, 0) {
        Ok(result) => println!("5/0 = {}", result),
        Err(e) => println!("Error: {}", e),
    }
}
